### PR TITLE
feat(runtime): support more than one VM.

### PR DESCRIPTION
**Pull request by olonho**
Added support for basic work with Wasmtime 0.17. Error handling and 
gas metering works the same way on all tests.

Test plan
---------

Made most contract test to execute with two VMs and then
cargo test --all

### CODE REVIEW

**Comment by SkidanovAlex** [comment]
That's a pretty lengthy change. Do we want to postpone it post-phase-I? (we already have a branch for such changes, it is called post-phase-1).

**Comment by olonho** [comment]
Yeah, we could, although if not implemented, phase-1 will be Wasmer only.

File: runtime/near-vm-runner/src/wasmtime_runner.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    fn fits_memory(&self, offset: u64, len: u64) -> bool {
+        match offset.checked_add(len) {
+            None => false,
+            Some(end) => self.0.size() as u64 >= end,
+        }
+    }
+
+    fn read_memory(&self, offset: u64, buffer: &mut [u8]) {
+        let offset = offset as usize;
+        unsafe {
```
Comment (frol) [inline runtime/near-vm-runner/src/wasmtime_runner.rs:41]: Is it enforced by MemoryLike trait contract that we can use `unsafe` here? Can we have comments around every `unsafe` usage on why it is safe and minimize the block to the specific unsafe expression, so it is easier to see which method is exactly unsafe?

File: runtime/near-vm-runner/src/wasmer_runner.rs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        Ok(x) => x,
+        Err(err) => return (None, Some(err)),
+    };
+    let mut memory = match WasmerMemory::new(
+        wasm_config.limit_config.initial_memory_pages,
+        wasm_config.limit_config.max_memory_pages,
+    ) {
+        Ok(x) => x,
+        Err(_err) => panic!("Cannot create memory for a contract call"),
+    };
```
Comment (frol) [inline runtime/near-vm-runner/src/wasmer_runner.rs:224]: You can replace this `match` with `.expect()` method.

File: runtime/near-vm-logic/src/logic.rs [hunk 3]
```diff
@@ -23,7 +23,7 @@ pub struct VMLogic<'a> {
     /// receipts creation.
     ext: &'a mut dyn External,
     /// Part of Context API and Economics API that was extracted from the receipt.
-    context: VMContext,
+    pub context: VMContext,
```
Comment (bowenwang1996) [inline runtime/near-vm-logic/src/logic.rs:26]: why is this necessary?

Comment (olonho) [reply runtime/near-vm-logic/src/logic.rs:26]: Oh, not needed at all, just debugging leftover, thanks for spotting.

**Comment by nearmax** [comment]
> That's a pretty lengthy change. Do we want to postpone it post-phase-I? (we already have a branch for such changes, it is called post-phase-1).

It adds a lot of code, but it is not invasive to our current code -- me and @olonho discussed hiding Wasmtime behind the compilation flag for now and not add it to the genesis config, and not expose the environment variable. We have several goals with this code:
* Verify Wasmer correctness by comparing it against Wasmtime in our tests -- something that is very important to do ASAP, because we had encountered Wasmer bugs in the past;
* Have Wasmtime as a backup runtime -- if we find some critical issue with Wasmer, we would like to be able to switch to Wasmtime without being delayed by R&D;
* In a long-run we want to run Wasmer and Wasmtime side-by-side in production, and if node encounters a situation when they disagree it should panic. This will add a significant level of safety to our runtime.

The first bullet point is something we would like to do ASAP, which we can achieve by hiding Wasmtime behind a compilation flag.

**Comment by SkidanovAlex** [comment]
Cool, I have no concerns pushing this to master.

File: runtime/near-vm-runner/src/wasmer_runner.rs (hunk 2 above)
Comment (olonho) [reply runtime/near-vm-runner/src/wasmer_runner.rs:214]: Sure, it was just moving existing code, although.

File: runtime/near-vm-runner/src/wasmtime_runner.rs (hunk 1 above)
Comment (olonho) [reply runtime/near-vm-runner/src/wasmtime_runner.rs:42]: Yeah, added comments.

File: runtime/near-vm-logic/src/config.rs [hunk 4]
```diff
@@ -1,8 +1,30 @@
...
+    /// Wasmer VM.
+    Wasmer,
+    /// Wasmtime VM.
+    Wasmtime,
+}
+
+impl Default for VMKind {
+    fn default() -> Self {
+        match env::var("NEAR_VM") {
+            Ok(val) => match val.as_ref() {
```
Comment (ailisp) [inline runtime/near-vm-logic/src/config.rs:18]: possibly simpler to be 
``` 
Ok("wasmtime") => VMKind::Wasmtime,
 _ =>VMKind::Wasmer
```

File: runtime/near-vm-runner/src/wasmtime_runner.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                VMError::FunctionCallError(FunctionCallError::HostError(h.clone()))
+            }
+            Some(VMLogicError::ExternalError(s)) => VMError::ExternalError(s.clone()),
+            Some(VMLogicError::InconsistentStateError(e)) => {
+                VMError::InconsistentStateError(e.clone())
+            }
+            None => panic!("Error is not properly set"),
+        }
+    } else {
+        // VMError::FunctionCallError(LinkError { msg: format!("{:#?}", trap) })
```
Comment (nearmax) [inline runtime/near-vm-runner/src/wasmtime_runner.rs:79]: Should we remove it here?

File: runtime/near-vm-runner/src/wasmtime_runner.rs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        }
+    }
+}
+
+impl IntoVMError for wasmtime::Trap {
+    fn into_vm_error(self) -> VMError {
+        if self.i32_exit_status() == Some(239) {
+            trap_to_error(&self)
+        } else {
+            // VMError::FunctionCallError(LinkError { msg: format!("{:#?}", self) })
```
Comment (nearmax) [inline runtime/near-vm-runner/src/wasmtime_runner.rs:100]: Should we remove it?

File: runtime/near-vm-runner/tests/test_error_cases.rs [hunk 7]
```diff
@@ -420,12 +476,58 @@ fn bad_many_imports() -> Vec<u8> {
...
+                vm_kind
+            ),
+            (
+                Some(vm_outcome_with_gas(100)),
+                Some(VMError::FunctionCallError(FunctionCallError::HostError(
+                    HostError::GasExceeded
+                )))
+            )
+        );
+    });
```
Comment (nearmax) [inline runtime/near-vm-runner/tests/test_error_cases.rs:532]: It would add safety if tests here were testing for all variants of `FunctionCallError` enum. It might go into a different PR.

File: runtime/near-vm-runner/src/wasmtime_runner.rs [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        }
+    } else {
+        // VMError::FunctionCallError(LinkError { msg: format!("{:#?}", trap) })
+        VMError::FunctionCallError(WasmUnknownError)
+    }
+}
+
+impl IntoVMError for anyhow::Error {
+    fn into_vm_error(self) -> VMError {
+        // TODO: incorrect
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/near-vm-runner/src/wasmtime_runner.rs:86]: Removed obsolete comment.

File: runtime/near-vm-runner/Cargo.toml [hunk 9]
```diff
@@ -16,6 +16,8 @@ This crate implements the specification of the interface that Near blockchain ex
 cached = "0.12"
 wasmer-runtime = { version = "=0.17.0", features = ["default-backend-singlepass"], default-features = false }
 wasmer-runtime-core = { version = "=0.17.0" }
+wasmtime = "0.17.0"
+anyhow = "1.0.19"
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/near-vm-runner/Cargo.toml:20]: This crate as needed, IMO, as otherwise I cannot implement IntoVMError for anyhow::Error, and this is an error returned by wasmtime::Linker::instantiate and function runner.

File: runtime/near-vm-logic/src/config.rs [hunk 10]
```diff
@@ -1,8 +1,30 @@
...
+pub enum VMKind {
+    /// Wasmer VM.
+    Wasmer,
+    /// Wasmtime VM.
+    Wasmtime,
+}
+
+impl Default for VMKind {
+    fn default() -> Self {
+        match env::var("NEAR_VM") {
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/near-vm-logic/src/config.rs:17]: Could you please elaborate why you find compilation option preferable here? IMO runtime flag gives us more flexibility, and I'd love to keep it, if possible.

File: runtime/near-vm-runner/src/wasmtime_runner.rs (hunk 5 above)
Comment (olonho) [reply runtime/near-vm-runner/src/wasmtime_runner.rs:79]: Done.

File: runtime/near-vm-runner/tests/test_error_cases.rs (hunk 7 above)
Comment (olonho) [reply runtime/near-vm-runner/tests/test_error_cases.rs:532]: Agree, we generally need to think about better coverage.

File: runtime/near-vm-runner/src/wasmtime_runner.rs (hunk 6 above)
Comment (olonho) [reply runtime/near-vm-runner/src/wasmtime_runner.rs]: Done.

File: runtime/near-vm-runner/Cargo.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-vm-runner/Cargo.toml:20]: Sounds bad that wasmtime doesn't wrap and export it's own error type

File: runtime/near-vm-logic/src/config.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-vm-logic/src/config.rs:17]: I think @nearmax means a combination of feature flag + env var. Compiled with both feature flag gives you ability to "override runtime with env var", but default feature flag: default_runtime_is_wasmer, another feature flag: default_runtime_is_wasmtime

File: runtime/near-vm-logic/src/config.rs (hunk 4 above)
Comment (olonho) [reply runtime/near-vm-logic/src/config.rs:18]: Unfortunately this doesn't work for me.

File: runtime/near-vm-logic/src/config.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (nearmax) [reply runtime/near-vm-logic/src/config.rs:17]: The problem is that we have validators that run our code and overall experiment with it. The general principle that we hold is that it should not be easy to run a node in a not-safe mode, just to avoid potential human error. Therefore various code parameters that are considered to be dangerous we hide behind compilation flags. In the case of wasmtime we don't want some validator to turn it on, because they think it is a safe configuration parameter, like timeout on our RPC, cache size, etc. There were cases in the blockchain world, even with some renowned blockchains, where validators would accidentally run a misconfigured node and get slashed, i.e. lose a lot of assets. Since wasmtime was not thoroughly studied by us, and it uses optimized compilation it is not safe to be used. We were even thinking going as far as protecting these kinds of features by three factors: compilation flag, command line argument, and special marker file -- meaning one would have to provide all three to enable it. Additionally, environment variables is not something that we use in nearcore, env variables make configuration dependencies hard to track (i.e. one cannot tell from the signature of the module's functions/structs what is the input, since env variable is a global statically accessible input).

File: runtime/near-vm-runner/Cargo.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (nearmax) [reply runtime/near-vm-runner/Cargo.toml:20]: I see, my original comment is probably wrong anyway, since wasmtime pulls in this dependency anyway.

File: runtime/near-vm-logic/src/config.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/near-vm-logic/src/config.rs]: Makes sense. Fixed the default selection to be compilation feature.

File: runtime/near-vm-runner/Cargo.toml [hunk 11]
```diff
@@ -16,6 +16,8 @@ This crate implements the specification of the interface that Near blockchain ex
 cached = "0.12"
 wasmer-runtime = { version = "=0.17.0", features = ["default-backend-singlepass"], default-features = false }
 wasmer-runtime-core = { version = "=0.17.0" }
+wasmtime = { version = "0.17.0", features = ["lightbeam"], default-features = false }
```
Comment (olonho) [inline runtime/near-vm-runner/Cargo.toml:19]: Also note, that I tried to use not so optimizing lightbeam backend.

File: runtime/near-vm-runner/src/runner.rs [hunk 12]
```diff
@@ -48,59 +24,57 @@ pub fn run<'a>(
...
 
-    match module.instantiate(&import_object) {
-        Ok(instance) => match instance.call(&method_name, &[]) {
-            Ok(_) => (Some(logic.outcome()), None),
-            Err(err) => (Some(logic.outcome()), Some(err.into_vm_error())),
-        },
-        Err(err) => (Some(logic.outcome()), Some(err.into_vm_error())),
+pub fn with_vm_variants(runner: fn(VMKind) -> ()) {
+    for vm in vec![VMKind::Wasmer, VMKind::Wasmtime].iter() {
+        runner(vm.clone());
```
Comment (frol) [inline runtime/near-vm-runner/src/runner.rs:78]: TIL that Rust implements array-value iterator only under a [nightly feature-gate "array_value_iter"](https://github.com/rust-lang/rust/issues/65798). However, in this case you can do the following:

1. Mark `VMKind` as `Copy` type (`#[derive(Copy)`)
2. ```rust
    for vm in &[VMKind::Wasmer, VMKind::Wasmtime] {
        runner(*vm);
    }
    ```

This way you don't need to allocate Vec.

Comment (olonho) [reply runtime/near-vm-runner/src/runner.rs:78]: Thanks!
